File: PowerOfN.java
// Time complexity: O(log(n))
// Space complexity: O(log(n))

// Approach: Calculate power of n/2; based on if n is even or odd
// adjust the result.

public class PowerOfN {
    public double myPow(double x, int n) {
        double res;

        if (n < 0) {
            res = raisedToPower(x, n * -1);
            return 1 / res;
        } else {
            return raisedToPower(x, n);
        }
    }

    // n >= 0
    public double raisedToPower(double x, int n) {
        if (n == 0) {
            return 1.0;
        }

        double halfPower = raisedToPower(x, n / 2);

        if (n % 2 == 0) {
            return halfPower * halfPower;
        } else {
            return halfPower * halfPower * x;
        }
    }
}

File: powerofN.py
class Solution:
    # Divide and Conquor: Time Complexity: log(n) # Space Complexity: O(1)
    def myPow(self, x: float, n: int) -> float:
        if(n==0):
            return 1
        if(n==1):
            return x
        if(n<1):
            x = 1/x
            n = n* -1
            
        
        def pow1(x,n):
            if(n==1):
                return x
            if(n==0):
                return 1
            
            if(n%2==0):
                num = pow1(x,n//2)
                return num*num
            else:
                num = pow1(x,(n-1)//2)
                return num*num*x
        
        return pow1(x,n)
